components/brave_shields/browser/custom_subscription_download_client.cc

```diff
@@ -0,0 +1,128 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_shields/browser/custom_subscription_download_client.h"
+
+#include <map>
+#include <memory>
+#include <set>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "base/bind.h"
+#include "base/metrics/histogram_macros_local.h"
+#include "base/threading/sequenced_task_runner_handle.h"
+#include "brave/components/brave_shields/browser/ad_block_service.h"
+#include "brave/components/brave_shields/browser/ad_block_subscription_service_manager.h"
+#include "brave/components/brave_shields/browser/custom_subscription_download_manager.h"
+#include "components/download/public/background_service/download_metadata.h"
+#include "services/network/public/cpp/resource_request_body.h"
+
+namespace brave_shields {
+
+CustomSubscriptionDownloadClient::CustomSubscriptionDownloadClient(
+    AdBlockSubscriptionServiceManager* subscription_manager)
+    : subscription_manager_(subscription_manager) {}
+
+CustomSubscriptionDownloadClient::~CustomSubscriptionDownloadClient() = default;
+
+CustomSubscriptionDownloadManager*
+CustomSubscriptionDownloadClient::GetCustomSubscriptionDownloadManager() {
+  return subscription_manager_->download_manager();
+}
+
+void CustomSubscriptionDownloadClient::OnServiceInitialized(
+    bool state_lost,
+    const std::vector<download::DownloadMetaData>& downloads) {
+  CustomSubscriptionDownloadManager* download_manager =
+      GetCustomSubscriptionDownloadManager();
+  if (!download_manager)
+    return;
+
+  std::set<std::string> outstanding_download_guids;
+  std::map<std::string, base::FilePath> successful_downloads;
+  for (const auto& download : downloads) {
+    if (!download.completion_info) {
+      outstanding_download_guids.emplace(download.guid);
+      continue;
+    }
+
+    successful_downloads.emplace(download.guid, download.completion_info->path);
+  }
+
+  download_manager->OnDownloadServiceReady(outstanding_download_guids,
+                                           successful_downloads);
+}
+
+void CustomSubscriptionDownloadClient::OnServiceUnavailable() {
+  CustomSubscriptionDownloadManager* download_manager =
+      GetCustomSubscriptionDownloadManager();
+  if (download_manager)
+    download_manager->OnDownloadServiceUnavailable();
+}
+
+void CustomSubscriptionDownloadClient::OnDownloadStarted(
+    const std::string& guid,
+    const std::vector<GURL>& url_chain,
+    const scoped_refptr<const net::HttpResponseHeaders>& headers) {
+  // Do not remove. This is a hook used by integration tests that test
```

Comment:
do we have any integration tests?

Reply:
Not specifically with this hook at the moment. I'll look at whether or not it makes sense to build out similar tests to what's in Chromium and either add those or just delete this.

Reply:
I couldn't find any tests that used that hook, so I just removed the method entirely. I do already have other tests that would fail if the download never completes successfully.